### PR TITLE
cli: duplicate: add setting for templating the new commit descriptions

### PR DESCRIPTION
**Pull request by winterqt**
This allows the customization of the duplicated commit descriptions.

An ideal use case for this is emulating `git cherry-pick -x`, as illustrated in the tests.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by winterqt** [comment]
Could probably use a mention in the Git command table at the very least (if we decide to keep this as a flag and not a setting a la `templates.draft_commit_description`).

File: cli/src/commands/duplicate.rs [hunk 1]
```diff
@@ -179,6 +199,48 @@ pub(crate) fn cmd_duplicate(
...
+                .map(|c| c.id().clone())
+                .collect(),
+            |rewriter| {
+                let old_dupe_commit_id = rewriter.old_commit().id().clone();
+                let commit_builder = rewriter.reparent();
+                if let Some((desc, old_commit_id)) = new_descs.remove(&old_dupe_commit_id) {
+                    *duplicated_commits
+                        .get_mut(&old_commit_id)
+                        .expect("duplicated_commits should contain the original commit id") =
+                        commit_builder.set_description(desc).write()?;
```
Comment (winterqt) [inline cli/src/commands/duplicate.rs:234]: Perhaps we should copy [`jj describe`'s new behavior](https://github.com/jj-vcs/jj/blob/5dd64ae978a20851151cddfe6b6421d929e70687/cli/src/commands/describe.rs#L264-L265) and set the commit timestamp to the old one? Sure, it'll be... a millisecond or two off, but... maybe it makes sense?

**Review by winterqt — COMMENTED** [review]
Excluding the efficiency of writing 2x the commits with this option enabled, the main downside of this method is that the evolog looks ugly with the extra commit (IMO):

```
    ○  yostqsxw test.user@example.com 2001-02-03 08:05:15 4037fb05
    │  a
    ○  yostqsxw hidden test.user@example.com 2001-02-03 08:05:15 3d341b2f
    │  a
    ○  rlvkpnrz test.user@example.com 2001-02-03 08:05:09 a 2443ea76
    │  a
    ○  rlvkpnrz hidden test.user@example.com 2001-02-03 08:05:08 d8d5f980
       (empty) a
```

However, the op log looks good. :)

```
    @  44a208b90324 test-username@host.example.com 2001-02-03 04:05:15.000 +07:00 - 2001-02-03 04:05:15.000 +07:00
    │  duplicate 1 commit(s)
    │  args: jj duplicate a -T 'concat(description, "\n", "(cherry-picked from ", commit_id, ")")'
    │
    │  Changed commits:
    │  ○  + yostqsxw 4037fb05 a
```

Probably not that big of a deal, though; our working copy snapshots put more "wear" on the repo than this ever could.

File: cli/tests/test_duplicate_command.rs [hunk 2]
```diff
@@ -2427,6 +2427,62 @@ fn test_rebase_duplicates() {
...
+    @    17a00fc21654   c
+    ├─╮
+    │ ○  d370aee184ba   b
+    ○ │  2443ea76b0b1   a
+    ├─╯
+    ◆  000000000000
+    [EOF]
+    ");
+
+    let template = r#"concat(description, "\n", "(cherry-picked from ", commit_id, ")")"#;
```
Comment (emilazy) [inline cli/tests/test_duplicate_command.rs:2451]:
```suggestion
    let template = r#"concat(description, "\n", "(cherry picked from ", commit_id, ")")"#;
```
to match Git.

File: cli/src/commands/duplicate.rs (hunk 1 above)
Comment (yuja) [inline cli/src/commands/duplicate.rs:234]: > the main downside of this method is that the evolog looks ugly with the extra commit

Perhaps, you can fix up `predecessors` to point to the original commit.

A better option is to split `duplicate_commits()` into two parts.

**Review by yuja — COMMENTED** [review]
The implementation generally looks good to me, thanks.

File: cli/src/commands/duplicate.rs (hunk 1 above)
Comment (winterqt) [reply cli/src/commands/duplicate.rs:234]: Yeah, it seems like splitting it may be the best bet. I tried to do some things with callbacks, to no avail, and I was thinking about splitting them, but then I need to think about how exactly to split the functions. I'll figure it out. :)

File: cli/src/config/templates.toml [hunk 3]
```diff
@@ -66,6 +66,8 @@ separate(" ",
 )
 '''
 
+duplicate_commit_description = ''
```
Comment (winterqt) [inline cli/src/config/templates.toml:69]: We probably don't want to add a default, but perhaps an alias called `cherry_pick_commit_description` or something would be good; then we can reference it in the Git command table.

Comment (winterqt) [reply cli/src/config/templates.toml:69]: Hm, should this be named `duplicate_description` instead, like `revert_description`?

File: lib/src/rewrite.rs [hunk 4]
```diff
@@ -776,6 +779,7 @@ pub struct DuplicateCommitsStats {
 pub fn duplicate_commits(
     mut_repo: &mut MutableRepo,
     target_commits: &[CommitId],
+    mut target_descriptions: HashMap<CommitId, String>,
```
Comment (yuja) [inline lib/src/rewrite.rs:782]: nit: pass by reference? I don't think the extra `.clone()` cost would matter.

File: cli/src/commands/duplicate.rs (hunk 1 above)
Comment (yuja) [reply cli/src/commands/duplicate.rs]: Right. Since the caller knows the commits to be duplicated, we don't have to split functions. (I thought there would be something that had to be computed.)

File: lib/src/rewrite.rs (hunk 4 above)
Comment (winterqt) [reply lib/src/rewrite.rs]: Sure, I suppose.

**Comment by winterqt** [comment]
Strange... it builds locally. I think Cargo is being dumb again. :/

File: cli/src/commands/duplicate.rs [hunk 5]
```diff
@@ -164,6 +170,26 @@ pub(crate) fn cmd_duplicate(
             }
         }
     }
+    let mut new_descs = HashMap::new();
+    if let Some(template) = tx
+        .settings()
+        .get_string("templates.duplicate_description")
+        .ok()
```
Comment (yuja) [inline cli/src/commands/duplicate.rs:177]: nit: `.get_string(..)?` to propagate error

File: cli/src/commands/duplicate.rs [hunk 6]
```diff
@@ -164,6 +170,26 @@ pub(crate) fn cmd_duplicate(
...
+        .filter(|t| !String::is_empty(t))
```
Comment (yuja) [inline cli/src/commands/duplicate.rs:178]: nit: `!t.is_empty()` or `t != "description"` (see the other comment)

File: cli/src/config/templates.toml [hunk 7]
```diff
@@ -66,6 +66,8 @@ separate(" ",
 )
 '''
 
+duplicate_description = ''
```
Comment (yuja) [inline cli/src/config/templates.toml:69]: Perhaps, the default should be `'description'`. The behavior shouldn't change depending on the optimization for the default template.

Can you also move this after `draft_commit_description`? Appears that the templates are sorted lexicographically.

File: lib/src/rewrite.rs [hunk 8]
```diff
@@ -766,8 +766,11 @@ pub struct DuplicateCommitsStats {
 /// The roots of `target_commits` are duplicated on top of the new
 /// `parent_commit_ids`, whilst other commits in `target_commits` are duplicated
 /// on top of the newly duplicated commits in the target set. If
-/// `children_commit_ids` is not empty, the `children_commit_ids` will be
-/// rebased onto the heads of the duplicated target commits.
+/// `target_descriptions` is not empty, it will be consulted to retrieve the new
+/// descriptions of the target commits, falling back to the original if the map
+/// does not contain an entry for a given commit. If `children_commit_ids` is
```
Comment (yuja) [inline lib/src/rewrite.rs:771]: nit: I would split this to new paragraph. Here we're discussing about the graph structure.

File: cli/tests/test_duplicate_command.rs [hunk 9]
```diff
@@ -2427,6 +2427,62 @@ fn test_rebase_duplicates() {
...
+    ├─╮
+    │ ○  d370aee184ba   b
+    ○ │  2443ea76b0b1   a
+    ├─╯
+    ◆  000000000000
+    [EOF]
+    ");
+
+    test_env.add_config(r#"templates.duplicate_description = "concat(description, '\n(cherry-picked from commit ', commit_id, ')')""#);
+    let output = work_dir.run_jj(["duplicate", "a"]);
```
Comment (yuja) [inline cli/tests/test_duplicate_command.rs:2452]: Can you add test for `duplicate_commits()`?

**Review by yuja — APPROVED** [review]
Thanks.

File: cli/src/commands/duplicate.rs (hunk 5 above)
Comment (winterqt) [reply cli/src/commands/duplicate.rs:177]: This was a leftover from when I didn't put it in the config by default, thanks.

File: cli/src/config/templates.toml (hunk 7 above)
Comment (winterqt) [reply cli/src/config/templates.toml:69]: > Perhaps, the default should be `'description'`. The behavior shouldn't change depending on the optimization for the default template.

Not sure what you mean -- the exact behavior wouldn't change at all due to the `is_empty` check. We're just changing the optimized case.

Also, to clarify: no matter if the template is `'description'` (optimized out) or `' description '`, the resulting commits will be identical.

File: cli/tests/test_duplicate_command.rs (hunk 9 above)
Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2452]: Let me know if this is sufficient.

File: lib/src/rewrite.rs [hunk 10]
```diff
@@ -776,6 +780,7 @@ pub struct DuplicateCommitsStats {
 pub fn duplicate_commits(
     mut_repo: &mut MutableRepo,
     target_commits: &[CommitId],
+    target_descriptions: &HashMap<CommitId, String>,
```
Comment (winterqt) [inline lib/src/rewrite.rs:783]: I chose not to make this `Into<String>` because you then have to use turbofishes for `HashMap::new()` (that is, the empty case).

File: cli/tests/cli-reference@.md.snap [hunk 11]
```diff
@@ -802,6 +802,8 @@ When none of the `--destination`, `--insert-after`, or `--insert-before` argumen
 
 When any of the `--destination`, `--insert-after`, or `--insert-before` arguments are provided, the roots of the specified commits will be duplicated onto the destination indicated by the arguments. Other specified commits will be duplicated onto these newly duplicated commits. If the `--insert-after` or `--insert-before` arguments are provided, the new children indicated by the arguments will be rebased onto the heads of the specified commits.
 
+The descriptions of the duplicated commits can be customized with the `templates.duplicate_description`.
```
Comment (winterqt) [inline cli/tests/cli-reference@.md.snap:805]: Maybe
```suggestion
By default, the duplicated commits retain the description of the originals. This can be customized with the `duplicate_description` template.
```

File: cli/src/config/templates.toml (hunk 7 above)
Comment (yuja) [reply cli/src/config/templates.toml]: > Not sure what you mean -- the exact behavior wouldn't change at all due to the `is_empty` check.

I mean the behavior shouldn't change if we removed the `if`. It should just exist for optimization purpose. Since an empty string is a valid template expression, unfortunately, it should be evaluated to `""`, not to `description`.

FWIW, if user defined `description` alias, the result would go awry. It might be safer to just always render the template.

File: cli/tests/test_duplicate_command.rs (hunk 9 above)
Comment (yuja) [reply cli/tests/test_duplicate_command.rs:2452]: I meant to add CLI test for both `duplicate_commits()` and `duplicate_commits_onto_parents()`, but adding lib tests seems also good.

File: cli/src/config/templates.toml (hunk 7 above)
Comment (winterqt) [reply cli/src/config/templates.toml]: Good point, thanks!

File: cli/tests/test_duplicate_command.rs (hunk 9 above)
Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2452]: What do you mean add CLI tests for them? `duplicate_commits()` is tested by the case I added (via `jj duplicate`), and I can add one for `duplicate_commits_onto_parents()` if you want.

It seems like you're talking about something different though -- could you please elaborate?

Comment (yuja) [reply cli/tests/test_duplicate_command.rs:2452]: `jj duplicate` (with no destination arguments) doesn't exercise the `duplicate_commits()` code path, right? For historical reasons, `jj duplicate` has two different behaviors.

Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2452]: Ah, right, so you meant that I currently test `duplicate_commits_onto_parents()`, but not `duplicate_parents()`. Yup, can fix. Sorry for the miscommunication, I thought you were talking about something else entirely.

File: cli/tests/cli-reference@.md.snap (hunk 11 above)
Comment (yuja) [reply cli/tests/cli-reference@.md.snap:805]: Looks good, but I wouldn't remove `templates.` prefix from the config name.

**Comment by winterqt** [comment]
Alright, this *should* be good to go now; would appreciate one more review pass. Thanks for the help!

File: cli/tests/cli-reference@.md.snap (hunk 11 above)
Comment (winterqt) [reply cli/tests/cli-reference@.md.snap:805]: https://jj-vcs.github.io/jj/latest/config/#default-description does, which is why I did.

Could do
```suggestion
By default, the duplicated commits retain the description of the originals. This can be customized with `templates.duplicate_description`.
```
but I would like them to be consistent.

File: cli/src/commands/duplicate.rs [hunk 12]
```diff
@@ -164,6 +170,26 @@ pub(crate) fn cmd_duplicate(
...
+        for commit_id in &to_duplicate {
+            let mut output = Vec::new();
+            let commit = tx.repo().store().get_commit(commit_id)?;
+            parsed
+                .format(&commit, &mut PlainTextFormatter::new(&mut output))
+                .expect("write() to vec backed formatter should never fail");
+
+            new_descs.insert(commit_id.clone(), output.into_string_lossy());
+        }
+    }
```
Comment (yuja) [inline cli/src/commands/duplicate.rs:191]: nit: (at your option), now we can remove `mut`, but I don't have preference over these.
```rust
let new_descs: HashMap<_, _> = {
    ...
    to_duplicate.iter().map(|commit_id| -> BackendResult<_> { .. }).try_collect()?
};

File: cli/tests/cli-reference@.md.snap (hunk 11 above)
Comment (yuja) [reply cli/tests/cli-reference@.md.snap:805]: In the command help context, it's not obvious where to put the `duplicate_description` config value. That's why I prefer fully-qualified name here.

**Review by yuja — APPROVED** [review]

File: cli/tests/test_duplicate_command.rs [hunk 13]
```diff
@@ -2427,6 +2427,96 @@ fn test_rebase_duplicates() {
...
+        a
+
+        (cherry-picked from commit 2443ea76b0b1c531326908326aab7020abab8e6c)
+
+    A a
+    [EOF]
+    ");
+
+    // Test duplicate_commits_onto_parents()
+    let output = work_dir.run_jj(["duplicate", "7e", "-B", "b8"]);
```
Comment (yuja) [inline cli/tests/test_duplicate_command.rs:2482]: nit: better to use bookmark name? you'll have to update hashes if something added to tests.

File: cli/src/commands/duplicate.rs (hunk 12 above)
Comment (winterqt) [reply cli/src/commands/duplicate.rs:191]: I'll think about it -- couldn't we do this before, though?

File: cli/tests/test_duplicate_command.rs (hunk 13 above)
Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2482]: Would it really be useful to use the bookmark names for some of the commits, but not others? (Only the ones from `create_commit` have the bookmarks set, and not the ones created by `jj duplicate`.

(I also think I ran into some issues when using the bookmark names for `create_commit` commits, but I can try again if you think it's worth it.)

File: cli/src/commands/duplicate.rs (hunk 12 above)
Comment (yuja) [reply cli/src/commands/duplicate.rs:191]: > couldn't we do this before, though?

You could, but we would need something like `new_descs = if { ... } else { HashMap::new() };` to remove `mut`.

File: cli/tests/test_duplicate_command.rs (hunk 13 above)
Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2482]: Sure, but we would still have to update the `jj show` commands for the newly duplicated commits. Do you think there's a better way to refer to those too?

Comment (yuja) [reply cli/tests/test_duplicate_command.rs:2482]: Maybe dump all commits at the end (by `jj log -T..`)?

btw, when we have to use hashes, I would suggest using 8-12 characters. The shortest hash is harder to spot, and has higher chance that it points to different commit by accident.

File: cli/tests/test_duplicate_command.rs [hunk 14]
```diff
@@ -2427,6 +2427,118 @@ fn test_rebase_duplicates() {
...
+        "template-aliases.description='\"alias\"'",
+    ]);
+    insta::assert_snapshot!(output, @r"
+    ------- stderr -------
+    Duplicated 3fb5c9d2839f as lylxulpl e6877e1e alias
+    [EOF]
+    ");
+
+    let template = r#"commit_id.short() ++ "\n" ++ description"#;
+    let output = work_dir.run_jj(["log", "-T", template]);
```
Comment (winterqt) [inline cli/tests/test_duplicate_command.rs:2517]: @yuja Do you think it's worthwhile to run `jj log` at every other commit if we're doing a big one at the end? Probably would be best to remove the intermediate ones.

Comment (yuja) [reply cli/tests/test_duplicate_command.rs:2517]: Yeah, I would also remove intermediate `jj log` calls so long as the result is obvious.

Comment (winterqt) [reply cli/tests/test_duplicate_command.rs:2517]: > so long as the result is obvious

Do you think that's true in this case? :)
